drivers/crypto/crypto_esp32_sha.c

```diff
@@ -0,0 +1,581 @@
+		}
+	} else {
+		for (size_t i = 0; i < state_len; i++) {
+			reg_addr_buf[i] = digest_state_words[i];
+		}
+	}
+
+	sha_ll_load(sha_type);
+}
+
+static inline void sha_hw_restore_esp32(struct esp_sha_ctx *s)
+{
+	if (s->first_block && s->algo != CRYPTO_HASH_ALGO_SHA224) {
+		return;
+	}
+
+	sha_ll_write_digest_esp32(s->hal_mode, s->H, s->state_words);
+}
+#endif /* SOC_SHA_SUPPORT_RESUME */
+
+static size_t sha_make_padding(const struct esp_sha_ctx *s, const uint8_t *tail, size_t tail_len,
+			       uint8_t last[128], uint8_t last2[128])
+{
+	const size_t B = s->block_bytes;
+	const size_t L = s->len_field_bytes;
+	const uint64_t bit_len = s->total_len * 8ULL;
+
+	if (tail_len >= B) {
+		LOG_ERR("Invalid tail length: %zu", tail_len);
+		return 0;
+	}
+
+	memset(last, 0, B);
+	if (tail_len > 0) {
+		memcpy(last, tail, tail_len);
+	}
+	last[tail_len] = 0x80;
+
+	if (tail_len + 1 + L <= B) {
+		if (L == 16) {
+			memset(&last[B - 16], 0, 8);
+		}
+
+		for (int i = 0; i < 8; i++) {
+			last[B - 8 + i] = (uint8_t)(bit_len >> (56 - 8 * i));
+		}
+		return 1;
+	}
+
+	memset(last2, 0, B);
+	if (L == 16) {
+		memset(&last2[B - 16], 0, 8);
+	}
+	for (int i = 0; i < 8; i++) {
+		last2[B - 8 + i] = (uint8_t)(bit_len >> (56 - 8 * i));
+	}
+
+	return 2;
+}
+
+static inline void sha_compress_block(const struct esp_sha_ctx *s, const uint8_t *block, bool first)
+{
+	const int words = s->block_bytes / 4;
+	uint32_t w[32];
+
+	for (int i = 0; i < words; i++) {
+		w[i] = sys_get_le32(&block[i * 4]);
+	}
+
+	bool first_param = (s->algo == CRYPTO_HASH_ALGO_SHA224) ? false : first;
+
+	sha_hal_hash_block(s->hal_mode, w, words, first_param);
+	sha_hal_wait_idle();
+}
+
+static int sha_update_stream(struct esp_sha_ctx *s, const uint8_t *in, size_t len)
+{
+	if (len == 0) {
+		return 0;
+	}
+
+	if (!in) {
+		return -EINVAL;
+	}
+
+	if (s->total_len + len < s->total_len) {
+		LOG_ERR("Total length overflow");
+		return -EOVERFLOW;
+	}
+
+	s->total_len += len;
+	uint8_t *buf = (uint8_t *)s->buf_w;
+
+	if (s->buf_len > 0) {
+		size_t take = MIN(len, s->block_bytes - s->buf_len);
+
+		memcpy(&buf[s->buf_len], in, take);
+		s->buf_len += take;
+		in += take;
+		len -= take;
+
+		if (s->buf_len == s->block_bytes) {
+			sha_compress_block(s, buf, s->first_block);
+			s->first_block = false;
+			s->buf_len = 0;
+		}
+	}
+
+	while (len >= s->block_bytes) {
+		sha_compress_block(s, in, s->first_block);
+		s->first_block = false;
+		in += s->block_bytes;
+		len -= s->block_bytes;
+	}
+
+	if (len > 0) {
+		memcpy(buf, in, len);
+		s->buf_len = len;
+	}
+
+	return 0;
+}
+
+static int sha_query_hw_caps(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+#if SOC_SHA_SUPPORT_RESUME
+	/* Chips with resume support can handle multi-part operations */
+	return CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS;
+#else
+	/* Original ESP32: No resume support, but still supports sync operations */
+	return CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS | CAP_NO_IV_PREFIX;
```

Comment:
Based on the documentation `CAP_NO_IV_PREFIX` is for block ciphers, not hash, but OTOH I cannot find any other flag to state "this driver doesn't support multipart operations". @ceolin should we add that?